DuggaSys/diagram.js

```diff
@@ -488,25 +508,38 @@ class StateMachine
         // If there is no history => return
         if (this.currentHistoryIndex == -1) return;
 
-        this.scrubHistory(this.currentHistoryIndex)
+        var time = this.historyLog[this.currentHistoryIndex - 1].time
+
+        do {
+            // Lower the historyIndex by one
+            this.currentHistoryIndex--;
+
+        }while(this.historyLog[this.currentHistoryIndex] && time == this.historyLog[this.currentHistoryIndex].time);
 
-        // Lower the historyIndex by one
-        this.currentHistoryIndex--;
+        this.scrubHistory(this.currentHistoryIndex);
 
         clearContext();
         showdata();
         updatepos(0, 0);
         displayMessage(messageTypes.SUCCESS, "Changes reverted!")
     }
-    stepForward(){
+    stepForward()
+    {
         // If there is not anything to restore => return
         if (this.historyLog.length == 0 || this.currentHistoryIndex == (this.historyLog.length -1)) return;
 
-        // Increase the currentHistoryIndex by one
-        this.currentHistoryIndex++;
+        // Go one step forward, if the next state in the history has the same time, do that too
+        do {
+
+            // Increase the currentHistoryIndex by one
+            this.currentHistoryIndex++;
 
-        // Restore the state
-        this.restoreState(this.historyLog[this.currentHistoryIndex]);
+            // Restore the state
+            this.restoreState(this.historyLog[this.currentHistoryIndex]);
+
+            var doNextState = false;
+            if (this.historyLog[this.currentHistoryIndex + 1]) doNextState = (this.historyLog[this.currentHistoryIndex].time == this.historyLog[this.currentHistoryIndex + 1].time)
```

Comment:
Break this down

Reply:
The selection is no longer a one-liner, bracket has been added to make it more readable.